pshtt/pshtt.py

```diff
@@ -525,6 +528,20 @@ def hsts_check(endpoint):
 
         header = endpoint.headers.get("Strict-Transport-Security")
 
+        # If there is no HSTS header, check the eventual response from redirects
+        if header is None and endpoint.ultimate_req and endpoint.url in endpoint.ultimate_req.url:
+            header = endpoint.ultimate_req.headers.get("Strict-Transport-Security")
+            if header:
+                logging.warning("{}: Found HSTS in redirected response from {}.".format(endpoint.url, endpoint.ultimate_req.url))
+
+        # If there is still no HSTS header, check all the redirect responses
+        if header is None and endpoint.ultimate_req and endpoint.ultimate_req.history:
+            for entry in endpoint.ultimate_req.history:
```

Comment:
I think the code here is assuming that the intermediate domains are somehow related to the initial domain.  What if the site redirects the user from `http://example.gov` to `http://www.example-2.gov` to `http://www.example-3.gov`?  If HSTS is present at the second or third sites, does that mean that we should give `http://example.gov` credit for being protected by HSTS?  It seems like the `http://example.gov` URL is still subject to hijack in that case, since the HSTS protection is associated with `example-2.gov` or `example-3.gov`.

Do we need to check a little deeper and make sure that the HSTS header would actually apply to the initial hostname?

I'll mark this as "needs changes" for now until we resolve this question.  I can undo that if I'm off the mark.

Reply:
I think this is still an open question.  I'd like to hear what @h-m-f-t thinks about this.

Reply:
@jsf9k Based on @h-m-f-t 's comment at https://github.com/cisagov/pshtt/issues/89#issuecomment-318242757 saying 

> I don't want compliance for compliance' sake, or maximal-compliance for a little security benefit— and the current approach does have a bit of that flavor. Happy to hear more discussion, but I think I'm supportive of making this change, especially given that SSL Labs does this.

I think @h-m-f-t is already supportive of making this change.

Reply:
@jsf9k The check for `endpoint.url in redirect_entry.url` in the `if` https://github.com/cisagov/pshtt/blob/0ec75ba3fcabffa43b5c265d92c3404fc77e0a56/pshtt/pshtt.py#L571 already ensures that the HSTS header only counts if it is coming from the initial endpoint/hostname, so in your example `http://example.gov` would not get credit.

Reply:
> @jsf9k The check for `endpoint.url in redirect_entry.url` in the `if`
> 
> https://github.com/cisagov/pshtt/blob/0ec75ba3fcabffa43b5c265d92c3404fc77e0a56/pshtt/pshtt.py#L571
> 
> already ensures that the HSTS header only counts if it is coming from the initial endpoint/hostname, so in your example `http://example.gov` would not get credit.

It gives credit if _any_ HSTS header encountered in the redirect chain involves the initial endpoint/hostname.  If the redirect chain goes on to send the user to an HTTP URL on an entirely different domain for which the browser has never seen an HSTS header, then it seems to me that the user is not being protected by the HSTS header.

Do you think this is a valid concern?  I'm OK deferring the resolution to #205, or even a separate issue.  I feel like there is a hole here, and if so then I'd like to at least document that before moving on.

Reply:
I think it is a valid concern, but I think it is a separate issue.  I think that the requirement for HSTS is only for the domain that we're checking, not for all endpoints/domains in the redirect chain.  Today we would give credit for HSTS to a domain if it returns a valid HSTS header in the initial response even if it then immediately redirects to an HTTP URL on an entirely different domain (although it should fail the separate Strictly Forces HTTPS check).  Here we are just continuing what we already do. 

I also think it is separate from https://github.com/cisagov/pshtt/issues/205 too. BOD 18-01 requires 
> All publicly accessible Federal websites and web services provide service through a secure connection

so I think it should fail some bigger check for a secure connection.  We already have a separate check for Domain Enforces HTTPS, so I think this check should be implemented there instead to check that there are no downgrades anywhere in the redirect chains for the domain's endpoints.  We could even add an additional check called Strictly Enforces HSTS to check that all endpoints in the redirect chain also have HSTS if you and the CyHy team think that should be a requirement.

Reply:
Created #206.